compiler/rustc_ast/src/token.rs

```diff
@@ -398,6 +398,30 @@ impl Token {
         }
     }
 
+    /// Returns `true` if the token can appear at the start of an pattern.
+    ///
+    /// Shamelessly borrowed from `can_begin_expr`, only used for diagnostics right now.
+    pub fn can_begin_pattern(&self) -> bool {
```

Comment:
I'm not totally confident this is totally accurate, but it's also only used for diagnostics for now...